src/shortcuts/__tests__/undo.ts

```diff
@@ -0,0 +1,295 @@
+import { importText } from '../../action-creators'
+import { MODALS, NOOP, RANKED_ROOT, ROOT_TOKEN } from '../../constants'
+import { exportContext } from '../../selectors'
+import { createTestStore } from '../../test-helpers/createTestStore'
+import { createMockStore } from '../../test-helpers/createMockStore'
+import executeShortcut from '../../test-helpers/executeShortcut'
+import undoShortcut from '../undo'
+import { initialState } from '../../util'
+import * as undoUtils from '../../util/isUndoEnabled'
+
+describe('undo shortcut', () => {
+  const event = { preventDefault: NOOP } as Event
+  const isUndoEnabled = jest.spyOn(undoUtils, 'isUndoEnabled')
+
+  it('dispatches undo action on shortcut if undo is enabled', () => {
+    // enable undo
+    isUndoEnabled.mockImplementationOnce(() => true)
+
+    const mockStore = createMockStore()
+    const store = mockStore(initialState())
+
+    executeShortcut(undoShortcut, { store, type: 'keyboard', event })
+
+    expect(store.getActions()).toEqual([{ type: 'undoAction' }])
+  })
+
+  it('does not dispatch an undo action if undo is disabled', () => {
+    // disable undo
+    isUndoEnabled.mockImplementationOnce(() => false)
+
+    const mockStore = createMockStore()
+    const store = mockStore(initialState())
+
+    executeShortcut(undoShortcut, { store, type: 'keyboard', event })
+
+    expect(store.getActions()).toEqual([])
+  })
+
+})
+
+it('undo thought change', async () => {
+
+  const store = createTestStore()
+
+  store.dispatch([
+    importText(RANKED_ROOT, `
+      - a
+      - b`
+    ),
+    { type: 'setCursor', thoughtsRanked: [{ value: 'a', rank: '0' }] },
+    {
+      type: 'existingThoughtChange',
+      newValue: 'aa',
+      oldValue: 'a',
+      context: [ROOT_TOKEN],
+      thoughtsRanked: [{ value: 'a', rank: 0 }]
+    },
+    { type: 'undoAction' }
+  ])
+
+  const exported = exportContext(store.getState(), [ROOT_TOKEN], 'text/plain')
+
+  const expectedOutput = `- ${ROOT_TOKEN}
+  - a
+  - b`
+
+  expect(exported).toEqual(expectedOutput)
+})
+
+it('group all navigation actions following an undoable(non-navigation) action and undo them together', async () => {
+
+  const store = createTestStore()
+
+  store.dispatch([
+    importText(RANKED_ROOT, `
+      - a
+      - b
+      - c
+      - d`
+    ),
+    { type: 'setCursor', thoughtsRanked: [{ value: 'b', rank: 1 }] },
+    { type: 'indent' },
+    {
+      type: 'existingThoughtChange',
+      newValue: 'b1',
+      oldValue: 'b',
+      context: ['a'],
+      rankInContext: 0,
+      thoughtsRanked: [{ value: 'a', rank: 0 }, { value: 'b', rank: 0 }]
+    },
+    { type: 'cursorBack' },
+    { type: 'moveThoughtDown' },
+    { type: 'cursorDown' },
+    // undo 'moveThoughtDown' and 'cursorDown'
+    { type: 'undoAction' }
+  ])
+
+  const cursorAfterFirstUndo = store.getState().cursor
+  expect(cursorAfterFirstUndo).toMatchObject([
+    { value: 'a' }])
```

Comment:
You'll need to assert the changed thought as well. You've only demonstrated that the `cursorDown` has been undone, not that `moveThoughtDown` has been undone in the same `undoAction`.